### PR TITLE
Improve type safety of the parser

### PR DESCRIPTION
**Pull request by rbuckton**
This follows on to #207 by making the parser type safe. This is a direct result of other work I've been doing to implement the Explicit Resource Management proposal into engine262.

The improved types found several potential issues, which have been addressed:
- `ExpressionParser.prototype.parseBracketedDefinition` was incorrectly using `isKeyword` when it should have been using `isKeywordRaw`.
- `ExpressionParser.prototype.parseLeftHandSideExpression` was incorrectly setting `node.arrowInfo.trailingComma` when it should have been setting `node.arrowInfo.hasTrailingComma`.

This PR opted to define specific parse nodes on a `ParseNode` namespace that merges with the base type, rather than introduce a large number of additional types as exports at the top level. If it is preferred that I should flatten the `ParseNode` namespace out to top-level types, I would be happy to do so.

To better capture how nodes are parsed, this PR introduces a somewhat complex `ParseNode.Unfinished<T>` type alias that is intended to capture both the actual state of a node as returned by `startNode()` as well as provide a modicum of type safety over the intended output types. In addition, a `ParseNode.Finished<T, K>` type alias that converts an `Unfinished` to its final type. The `Unfinished` type could be removed, but would require a significant number of complex casts and other refactors throughout the rest of the parser.

To provide better type safety for `Token` and related values in _tokens.mts_, there are a number of other somewhat complex type aliases whose intent is to preserve the exact types for the values produced in that file, without changing the actual underlying runtime code. An alternative to these complex types would be to refactor _tokens.mts_ to use more structured declaration forms like `enum`, but this would require a comprehensive rewrite of the entire file.

### CODE REVIEW

File: src/parser/ParseNode.mts
Comment (rbuckton) [reply src/parser/ParseNode.mts:15]: I considered this, but the problem is that such a large union results in significant slowdowns in check time when combined with control flow analysis. This causes significant typing and completion delays in an editor and is one of the reasons TypeScript doesn't do this in our own compiler yet.

File: src/parser/tokens.mts
Comment (rbuckton) [reply src/parser/tokens.mts:203]: It's asserting the exact operation the expression body of the function is testing. Essentially it says that `t` is one of the values in `KeywordLookup`. All of these types are constructed so as to maintain the existing runtime behavior. Using `enum` would make things far clearer in the type system, but would require a complete rewrite of the file (as I mentioned in the PR description).

File: src/parser/ParseNode.mts
Comment (rbuckton) [reply src/parser/ParseNode.mts:4]: That's certainly feasible. I was primarily focused on getting the correct intellisense and checks so that I could make progress on parser support for `using`, and less focused on the mutability of the types.

File: src/parser/ParseNode.mts
Comment (devsnek) [reply src/parser/ParseNode.mts:15]: personally i prefer the union type. anecdotally we have pretty large unions (500+ variants) in our codebase at work and it doesn't seem to behave too poorly.

**Comment by devsnek** [comment]
> The improved types found several potential issues, which have been addressed:

just out of curiosity did this reveal any missing behavior we could add to test262?

File: src/parser/BaseParser.mts
Comment (rbuckton) [reply src/parser/BaseParser.mts:15]: Not precisely. It's repurposing a ParseNode to be used as something else, which is prevalent throughout the parser, but doing so in a way that is a bit easier for TypeScript to work with. It's not necessary for a cover grammar to reuse the exact same object references, only that it reuses the same parse information. 

engine262 could just as easily just create a new ParseNode and copy the relevant information to it which actually might be preferrable in the long run. I'm not sure if engine262 is concerned about runtime performance, but the various `delete` operations that are performed when repurposing nodes are a major performance concern. `delete` takes an object out of "fast mode" and puts it into "dictionary mode". That can wreak havoc with inline caches and significantly increase polymorphism.

**Comment by rbuckton** [comment]
> > The improved types found several potential issues, which have been addressed:
> 
> just out of curiosity did this reveal any missing behavior we could add to test262?

No, but I have found a few places where engine262 doesn't match the specification, though it is in other code I added types to locally that builds on what is in this PR. I'm trying to break these changes up into more manageable chunks, and having a well-typed `ParseNode` and its subtypes is necessary for any other type annotations I might introduce.

One example of such a discrepancy was in `NewGlobalEnvironment()`. In Step 2 you have

```js
// 2. Let dclRec be a new declarative Environment Record containing no bindings.
const dclRec = new DeclarativeEnvironmentRecord(Value.null);
```

But the current spec is

```
2. Let dclRec be NewDeclarativeEnvironment().
```

The `NewDeclarativeEnvironment()` AO has additional initialization logic that isn't handled currently, which mattered for `using` because that's where I set up the `[[DisposeCapability]]` internal slot.

I plan to include that fix in the actual `using` PR, since it doesn't seem to affect engine262 currently.

**Comment by rbuckton** [comment]
> > The improved types found several potential issues, which have been addressed:
> 
> just out of curiosity did this reveal any missing behavior we could add to test262?

I did find a few places where the use of `raiseEarly` results in a parse tree that isn't strictly legal per the ECMA262 AST. Those specific cases are called out in `ParseNode.mts` where it says something like `// NOTE: Should be ...`, such as in [`NewExpression`](https://github.com/engine262/engine262/pull/215/files?show-viewed-files=true&file-filters%5B%5D=#diff-434db345ed5f5f59928c7d5422693ed9c81563d25a53f81bc315f0b231af9abdR763-R764). I chose to be more permissive with the types for now so as to not introduce too much in the way of actual changes to runtime logic.

File: src/parser/ParseNode.mts
Comment (rbuckton) [reply src/parser/ParseNode.mts:15]: It's less about the union itself, and more about the number of places where you have `if` and `switch` statements that select amongst discriminants as part of control flow. Almost every TS release includes some change to improve this, but it is a regular performance issue we see in large code bases that use React and other libraries that make heavy use of large unions. 

As part of TypeScript's regular benchmarking we have two versions of the Compiler we test: the compiler as-is, though from several versions ago, and "compiler with unions", which was a rewrite of the compiler to use a large, discriminated union for `Node`. The "compiler with unions" scenario is still far slower, but is something we are keeping an eye on.

File: src/parser/ParseNode.mts
Comment (Jack-Works) [reply src/parser/ParseNode.mts:4]: Yes, we can add that later. I always wanted a stricter version of TypeScript that things are immutable by default and use something like `writable` to opt-in.

This won't block the PR but if we have it that will be awesome.

File: src/parser/ParseNode.mts
Comment (Jack-Works) [reply src/parser/ParseNode.mts:15]: We can try union first. Engine262 is not as big as TypeScript, it does not have a checker/emitter/transformer.

File: src/parser/ParseNode.mts
Comment (rbuckton) [reply src/parser/ParseNode.mts]: I've made all nodes readonly by default, and mutable when passed through `Unfinished`.

Comment (rbuckton) [reply src/parser/ParseNode.mts]: I've also made the various array types readonly as well.

File: src/parser/ParseNode.mts
Comment (Jack-Works) [inline src/parser/ParseNode.mts:874]: I wonder what this `OrHigher` exactly means, I know it is related to the syntax "hierarchy". I saw it in the TypeScript compiler too.

Comment (rbuckton) [reply src/parser/ParseNode.mts:874]: It means "this node, or a node of a higher precedence".

**Comment by Jack-Works** [comment]
This PR contains a lot, I'll continue to review another file later, thanks!

File: src/parser/ParseNode.mts
Comment (rbuckton) [reply src/parser/ParseNode.mts:264]: They are already represented as one node in the parser. I was trying to cut down on unnecessary additional changes in the parser when possible. I imagine refactoring can happen in a later PR.

File: src/parser/ParseNode.mts
Comment (rbuckton) [reply src/parser/ParseNode.mts:434]: It is not used by the parser as it already makes a determination during parse whether it would be needed.

File: src/parser/ParseNode.mts
Comment (rbuckton) [reply src/parser/ParseNode.mts:725]: `#x in y`

File: src/parser/ParseNode.mts
Comment (rbuckton) [reply src/parser/ParseNode.mts:933]: What is confusing? This merely illustrates the precedence of the comma operator in relation to assignment.

File: src/parser/ParseNode.mts
Comment (rbuckton) [reply src/parser/ParseNode.mts:1350]: I would prefer it be in a later PR, possibly after the resource management PR is up and merged, if possible. Changing it now would complicate that work due to `for (using x of y)`.

File: src/parser/ParseNode.mts
Comment (rbuckton) [reply src/parser/ParseNode.mts:255]: I had concidered `PropertyDefinitionOrHigher`, but it isn't really a precedence thing.

File: src/parser/ParseNode.mts
Comment (Jack-Works) [reply src/parser/ParseNode.mts:725]: is PrivateIdentifier a RelationalExpression? according to the production, it looks like it should be in the `readonly RelationalExpression?: RelationalExpressionOrHigher;`

File: src/parser/ParseNode.mts
Comment (Jack-Works) [reply src/parser/ParseNode.mts:933]: I'll open an issue for this in the ECMA262 spec repo

File: src/parser/ParseNode.mts
Comment (rbuckton) [reply src/parser/ParseNode.mts:725]: No, it isn't. PrivateIdentifier is only allowed on its own when on the left side of an `in` expression, so it can't be a RelationalExpression on its own. Also, this type is reflecting the current behavior of the parser.

Comment (Jack-Works) [reply src/parser/ParseNode.mts:725]: oh I understand. Your comment is missing the [PrivateIdentifier](https://tc39.es/ecma262/multipage/ecmascript-language-lexical-grammar.html#prod-PrivateIdentifier) in [ShiftExpression](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#prod-ShiftExpression) production. I'll add it for you

**Review by Jack-Works — APPROVED** [review]

**Comment by Jack-Works** [comment]
looks good to me! @devsnek do you want to review this in detail or can we merge this?

File: src/parser/ExpressionParser.mts
Comment (rbuckton) [inline src/parser/ExpressionParser.mts:1467]: I'd been thinking of handling this using generics and callbacks to avoid the unsound cast.

File: src/parser/ExpressionParser.mts
Comment (rbuckton) [inline src/parser/ExpressionParser.mts:1468]: Why not turn the boolean pair into a `FunctionKind` rather than repeating it?

File: src/parser/FunctionParser.mts
Comment (rbuckton) [inline src/parser/FunctionParser.mts:30]: This can't be readonly as directives are written into it.

Comment (rbuckton) [reply src/parser/FunctionParser.mts:30]: To clarify, the implementation in `StatementParser` takes a writable array, not a `readonly` one. The fact this isn't an error is a bug in the compiler.

File: src/parser/ExpressionParser.mts
Comment (rbuckton) [reply src/parser/ExpressionParser.mts:1468]: Ah, never mind. `FunctionKind` only currently indicates sync or async functions.

File: src/parser/ExpressionParser.mts
Comment (rbuckton) [inline src/parser/ExpressionParser.mts]: I have mixed feelings about eslint's `no-nested-ternaries` rule. I think ternaries nested in the *true* branch are hard to read, but ternaries nested in the *false* branch are perfectly fine, especially when separated by line breaks, e.g.:

```js
const name =
  isAsync && isGenerator ? 'AsyncGeneratorMethod' :
  isAsync ? 'AsyncMethod' :
  isGenerator ? 'GeneratorMethod' :
  'MethodDefinition'
```
It makes me wonder whether eslint would forbid `match` expressions considering they will be laid out similarly.

Comment (rbuckton) [reply src/parser/ExpressionParser.mts]: I also don't like that it introduces a `let` where a `const` would be better.

**Comment by rbuckton** [comment]
Is type checking performed during CI? It doesn't seem to be, since rollup isn't report TS type checking errors. Should I add a `- run: npm run check` step to .github\workflows\test.yml?

File: src/parser/FunctionParser.mts
Comment (rbuckton) [reply src/parser/FunctionParser.mts:35]: I reverted this line.

File: src/parser/ParseNode.mts
Comment (rbuckton) [reply src/parser/ParseNode.mts:302]: No it is not, but the parser uses `parseInitializerOpt()` currently: https://github.com/engine262/engine262/pull/215/files?show-viewed-files=true&file-filters%5B%5D=#diff-0bbecd15b7c32e9995a70c92632b90c071e87c0e7bf939bacbf475f205eae201R1406. For the most part I'd prefer to keep spec conformance issues out of this PR except where entirely necessary.

File: src/parser/FunctionParser.mts
Comment (Jack-Works) [reply src/parser/FunctionParser.mts:35]: oh, when I change it to readonly the compiler doesn't give an error so I kept that.

File: src/parser/ExpressionParser.mts
Comment (Jack-Works) [reply src/parser/ExpressionParser.mts]: I have no idea about this rule. but I agree that nested ternary without parentheses is hard to read even if it is "nested in the false branch". that's because I need to confirm if it is _nested in the false branch_.

**Comment by Jack-Works** [comment]
> Is type checking performed during CI? It doesn't seem to be, since rollup isn't report TS type checking errors. Should I add a `- run: npm run check` step to .github\workflows\test.yml?

Not yet because we don't have enough files that passes the type check before.

I think you can add one.

File: src/parser/ExpressionParser.mts
Comment (devsnek) [reply src/parser/ExpressionParser.mts]: lol i find that code block to be very hard to read. i'd vote for keeping the rule.

**Comment by devsnek** [comment]
> looks good to me! @devsnek do you want to review this in detail or can we merge this?

i don't think i'll find time to review this in detail so i'd say go for it

**Pull request merged**
